core/rpc/src/author/mod.rs

```diff
@@ -104,25 +111,36 @@ impl<B, E, P, RA> AuthorApi<ExHash<P>, BlockHash<P>> for Author<B, E, P, RA> whe
 	type Metadata = crate::metadata::Metadata;
 
 	fn submit_extrinsic(&self, ext: Bytes) -> Result<ExHash<P>> {
-		let xt = Decode::decode(&mut &ext[..]).ok_or(error::Error::from(error::ErrorKind::BadFormat))?;
+		let xt = <Extrinsic<P>>::decode(&mut &ext[..]).ok_or(error::Error::from(error::ErrorKind::BadFormat))?;
 		let best_block_hash = self.client.info()?.chain.best_hash;
-		self.pool
-			.submit_one(&generic::BlockId::hash(best_block_hash), xt)
-			.map_err(|e| e.into_pool_error()
-				.map(Into::into)
-				.unwrap_or_else(|e| error::ErrorKind::Verification(Box::new(e)).into())
-			)
+		let res = self.transaction_pool
+			.submit_one(&generic::BlockId::hash(best_block_hash), xt.clone())
+			.map_err(|e| e.into_pool_error());
+
+		match res {
+			Ok(hash) => Ok(hash),
+			Err(Ok(pool_error)) => match (&self.inherent_pool, pool_error.kind()) {
+				(&Some(ref inherent_pool), &transaction_pool::txpool::error::ErrorKind::InvalidTransaction(_)) => {
+					warn!("Submitted extrinsic doesn't look like a valid transaction ({:?}). Assuming it's inherent.", pool_error);
+					let hash = self.transaction_pool.hash_of(&xt);
+					inherent_pool.add(xt);
+					Ok(hash)
+				},
+				_ => Err(pool_error.into()),
+			},
+			Err(Err(e)) => Err(error::ErrorKind::Verification(Box::new(e)).into()),
+		}
 	}
 
 	fn pending_extrinsics(&self) -> Result<Vec<Bytes>> {
-		Ok(self.pool.ready().map(|tx| tx.data.encode().into()).collect())
+		Ok(self.transaction_pool.ready().map(|tx| tx.data.encode().into()).collect())
```

Comment:
how do the inherents actually get into authoring? we probably want to `chain` this with the `inherents_pool`?

The behavior I'm expecting is
  - transactions are submitted and propagated over network
  - inherents are submitted but not propagated
  - both are included (but inherents first)? 

It's a bit weird to me because I would think that most inherents need to be localized to a specific block. I would guess that the only reason the UTXO chain is using that is because #2181 doesn't exist and we are using inherents as a way for the consensus engine to issue transactions.

Also it can really fuck shit up to include random inherents so we better watch out on that :P

Reply:
> how do the inherents actually get into authoring? we probably want to chain this with the inherents_pool?

Indeed the block is constructed as follows:
1. Inherent data generated by modules
2. Stuff from inherent pool (https://github.com/paritytech/substrate/blob/81993e8c2eaaf354b3b772114a7e1bfda795fbd8/core/basic-authorship/src/basic_authorship.rs#L210)
3. Transactions from the transaction pool.

> It's a bit weird to me because I would think that most inherents need to be localized to a specific block. I would guess that the only reason the UTXO chain is using that...

The main confusion was is in naming. UTXO uses "unsigned transactions" - they are still ok to be propagated, pass `validate_transaction` from the runtime we are just not able to tell (a single) account sender/index for each. Other than being "unsigned" they have not much in common with other inherent extrinsics that are just added by block producers.
I think eventually it turned out that no RPC changes were required for the UTXO chain.

> Also it can really fuck shit up to include random inherents so we better watch out on that :P

Indeed, the same issue arises when running offchain workers (that's what brought `InherentsPool` to existence in the first place), my undersanding is that the worst thing that will happen is the block will be rejected by the rest of the network, no? I hope all runtimes are written in a way that prevents bricking them :)

Reply:
> the worst thing that will happen is the block will be rejected by the rest of the network

if we are sending them around, we need a protection at the transaction-queue or authoring level. Otherwise it will be possible that none of the authorities can author a valid block.

Reply:
> if we are sending them around, we need a protection at the transaction-queue or authoring level.

Inherents are not propagated over the network and my last sentence referred to locally submitted inherents. The protections for stuff that actually ends up in the transaction queue (and is propagated) is implemented within the runtime as `validate_transaction` function.